single_kernel_mongo/managers/ldap.py

```diff
@@ -0,0 +1,175 @@
+
+from single_kernel_mongo.config.literals import (
+    Substrates,
+    TrustStoreFiles,
+)
+from single_kernel_mongo.config.relations import ExternalRequirerRelations
+from single_kernel_mongo.core.status_provider import StatusProvider
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.exceptions import (
+    DeferrableFailedHookChecksError,
+    LDAPSNotEnabledError,
+    NonDeferrableFailedHookChecksError,
+    WaitingForLdapDataError,
+)
+from single_kernel_mongo.lib.charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateTransferRequires,
+)
+from single_kernel_mongo.lib.charms.glauth_k8s.v0.ldap import LdapRequirer
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.core.operator import OperatorProtocol
+
+logger = getLogger(__name__)
+
+CANNOT_INTEGRATE_WITH_SHARD_STATUS = BlockedStatus("Cannot integrate LDAP with shard.")
+
+
+class LDAPManager(Object, StatusProvider):
+    """Manages the relation between glauth-k8s and replica set, config-sever or mongos router."""
+
+    def __init__(
+        self,
+        dependent: OperatorProtocol,
+        state: CharmState,
+        substrate: Substrates,
+        relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP,
+        cert_relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP_CERT,
+    ):
+        super().__init__(parent=dependent, key=relation_name)
+        self.dependent = dependent
+        self.charm = dependent.charm
+        self.workload = self.dependent.workload
+        self.state = state
+        self.substrate = substrate
+        self.relation_name = relation_name
+        self.cert_relation_name = cert_relation_name
+        self.ldap_requirer = LdapRequirer(self.charm, self.relation_name)
+        self.certificate_transfer = CertificateTransferRequires(self.charm, self.cert_relation_name)
+
+    def assert_pass_hook_checks(self) -> None:
+        """Runs some hook checks before allowing the hook to run."""
+        if not self.state.db_initialised:
+            raise DeferrableFailedHookChecksError("DB is not initialised")
+        if self.state.is_role(MongoDBRoles.SHARD):
+            self.charm.status_manager.set_and_share_status(CANNOT_INTEGRATE_WITH_SHARD_STATUS)
+            raise NonDeferrableFailedHookChecksError("Cannot integrate LDAP with shard.")
+        if self.state.upgrade_in_progress:
+            raise DeferrableFailedHookChecksError(
+                "Adding LDAP is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+        return
+
+    def is_valid_ldap_integration(self) -> bool:
+        """Returns True if the integration to ldap is valid."""
+        return (self.state.ldap_relation is not None) and not self.state.is_role(MongoDBRoles.SHARD)
+
+    def on_ldap_ready(self, relation: Relation) -> None:
```

Comment:
Could you rename all of these methods to a more "manager" friendly name, that indicates the actual business logic it is handling. The `on_*` naming should be reserved to the events